extensions/arc/runtime/src/main/java/io/quarkus/arc/runtime/context/ArcContextProvider.java

```diff
@@ -49,17 +41,31 @@ public ThreadContextSnapshot currentContext(Map<String, String> map) {
                 // context active, store current state, feed it new one and restore state afterwards
                 InjectableContext.ContextState stateToRestore = requestContext.getState();
                 requestContext.deactivate();
-                requestContext.activate(state);
+                if (state != null) {
+                    // only activate if previous thread had it active
+                    requestContext.activate(state);
+                }
                 controller = () -> {
                     // clean up, reactivate context with previous values
+                    if (state != null) {
+                        // only deactivate if previous thread had it active
+                        requestContext.deactivate();
+                    }
                     requestContext.deactivate();
```

Comment:
Doesn't that deactivate it twice?

Reply:
That's an oopsie, the second one shouldn't be there, thanks!